### PR TITLE
Fix telegram-bot rate limit

### PR DESCRIPTION
**Pull request by MAH69IK**
##### Summary
When sending too many messages in telegram bot - we can encounter rate limit and get 429 error. As a result, the message will be lost. With this proposal we'll never lose a message but try send it again after 1 second until this message could be sended or another error will appear.

FAQ about telegram limits: https://core.telegram.org/bots/faq#my-bot-is-hitting-limits-how-do-i-avoid-this

### CODE REVIEW

File: health/notifications/alarm-notify.sh.in [hunk 1]
```diff
@@ -1335,21 +1335,31 @@ send_telegram() {
...
+        successful=1
+
+        if [ "${httpcode}" = "200" ]; then
+          info "sent telegram notification for: ${host} ${chart}.${name} is ${status} to '${chatid}'"
+          sent=$((sent + 1))
+        elif [ "${httpcode}" = "401" ]; then
+          error "failed to send telegram notification for: ${host} ${chart}.${name} is ${status} to '${chatid}': Wrong bot token."
+        elif [ "${httpcode}" = "429" ]; then
+          error "failed to send telegram notification for: ${host} ${chart}.${name} is ${status} to '${chatid}': rate limit exceeded, retrying after 1s."
+          successful=0
```
Comment (thiagoftsm) [inline health/notifications/alarm-notify.sh.in:1357]: Hello @MAH69IK ,

Firstly thank you for your contribution!
I tested your PR and I observed that it is working as expected, but I have a concern with a border case. If some how we begin to receive only `429` we will enter in an infinite loop here, considering this possibility I will suggest to you also add a counter here creating a limit of tries for this modification you are bringing for us.

Best regards!

**Comment by MAH69IK** [comment]
Hi @thiagoftsm ! I added a counter to while loop. I chose the limit of 60 attempts because with 1 second delay - this provides new minute and new rate limit on the Telegram side (although in my tests I did not go beyond 32 attempts).

**Comment by thiagoftsm** [comment]
> Hi @thiagoftsm ! I added a counter to while loop. I chose the limit of 60 attempts because with 1 second delay - this provides new minute and new rate limit on the Telegram side (although in my tests I did not go beyond 32 attempts).

Thank you very much. I will review your PR today yet, I have another two in front of this, but it was already added to the queue. :)

**Review by thiagoftsm — APPROVED** [review]
I tested this PR with following alert : 

```conf
 alarm: dev_dim_template
    on: system.cpu
    os: linux
lookup: sum -3s at 0 every 3 percentage foreach *
 units: %
 every: 1s
  warn: $this > 1
  crit: $this > 4
````

To generate a great sample of alerts and observe `netdata` behavior with this PR. I observed that thanks it I had different tries and I got the expected alerts:

```sh
2022-06-23 02:38:23: alarm-notify.sh: INFO: sent telegram notification for: hades system.cpu.dev_dim_template_system is CRITICAL to '-XXXXXXXX'
2022-06-23 02:38:23: alarm-notify.sh: INFO: sent telegram notification for: hades system.cpu.dev_dim_template_iowait is CLEAR to '-XXXXXXXX'
2022-06-23 02:38:25: alarm-notify.sh: ERROR: failed to send telegram notification for: hades system.cpu.dev_dim_template_iowait is WARNING to '-XXXXXXXX': rate limit exceeded, retrying after 1s.
2022-06-23 02:38:28: alarm-notify.sh: ERROR: failed to send telegram notification for: hades system.cpu.dev_dim_template_iowait is WARNING to '-XXXXXXXX': rate limit exceeded, retrying after 1s.
2022-06-23 02:38:30: alarm-notify.sh: ERROR: failed to send telegram notification for: hades system.cpu.dev_dim_template_iowait is WARNING to '-XXXXXXXX': rate limit exceeded, retrying after 1s.
2022-06-23 02:38:31: alarm-notify.sh: ERROR: failed to send telegram notification for: hades system.cpu.dev_dim_template_iowait is WARNING to '-XXXXXXXX': rate limit exceeded, retrying after 1s.
2022-06-23 02:38:33: alarm-notify.sh: ERROR: failed to send telegram notification for: hades system.cpu.dev_dim_template_iowait is WARNING to '-XXXXXXXX': rate limit exceeded, retrying after 1s.
2022-06-23 02:38:35: alarm-notify.sh: ERROR: failed to send telegram notification for: hades system.cpu.dev_dim_template_iowait is WARNING to '-XXXXXXXX': rate limit exceeded, retrying after 1s.
2022-06-23 02:38:37: alarm-notify.sh: ERROR: failed to send telegram notification for: hades system.cpu.dev_dim_template_iowait is WARNING to '-XXXXXXXX': rate limit exceeded, retrying after 1s.
2022-06-23 02:38:39: alarm-notify.sh: ERROR: failed to send telegram notification for: hades system.cpu.dev_dim_template_iowait is WARNING to '-XXXXXXXX': rate limit exceeded, retrying after 1s.
2022-06-23 02:38:41: alarm-notify.sh: ERROR: failed to send telegram notification for: hades system.cpu.dev_dim_template_iowait is WARNING to '-XXXXXXXX': rate limit exceeded, retrying after 1s.
2022-06-23 02:38:43: alarm-notify.sh: ERROR: failed to send telegram notification for: hades system.cpu.dev_dim_template_iowait is WARNING to '-XXXXXXXX': rate limit exceeded, retrying after 1s.
2022-06-23 02:38:45: alarm-notify.sh: ERROR: failed to send telegram notification for: hades system.cpu.dev_dim_template_iowait is WARNING to '-XXXXXXXX': rate limit exceeded, retrying after 1s.
2022-06-23 02:38:47: alarm-notify.sh: ERROR: failed to send telegram notification for: hades system.cpu.dev_dim_template_iowait is WARNING to '-XXXXXXXX': rate limit exceeded, retrying after 1s.
2022-06-23 02:38:56: alarm-notify.sh: ERROR: failed to send telegram notification for: hades system.cpu.dev_dim_template_iowait is WARNING to '-XXXXXXXX': rate limit exceeded, retrying after 1s.
2022-06-23 02:38:58: alarm-notify.sh: INFO: sent telegram notification for: hades system.cpu.dev_dim_template_iowait is WARNING to '-XXXXXXXX'
```

Thank you very much @MAH69IK ! LGTM!

File: health/notifications/alarm-notify.sh.in [hunk 2]
```diff
@@ -1335,21 +1335,33 @@ send_telegram() {
...
-        error "failed to send telegram notification for: ${host} ${chart}.${name} is ${status} to '${chatid}': Wrong bot token."
-      else
-        error "failed to send telegram notification for: ${host} ${chart}.${name} is ${status} to '${chatid}' with HTTP response status code ${httpcode}."
-      fi
+      successful=0
+      attempt_counter=0
+
+      while [ ${successful} -eq 0 ] && [ ${attempt_counter} -lt 60 ]; do
+        # https://core.telegram.org/bots/api#sendmessage
+        httpcode=$(docurl ${disableNotification} \
```
Comment (ilyam8) [inline health/notifications/alarm-notify.sh.in:1343]: @thiagoftsm what Netdata processes are blocked while we are in this loop?

**Review by ilyam8 — APPROVED** [review]

**Comment by ilyam8** [comment]
@MrZammler highlighted the fact that health retries can delay the Agent shutdown

https://github.com/netdata/netdata/blob/3a96ef03b97c885122a4950733b8b01cb0fac69e/health/health.c#L1166-L1172

I am converting this PR to draft so it doesn't get merged until we discuss ^^

**Comment by MrZammler** [comment]
I've added a 20 seconds sleep in `alarm-notify.sh` script, and tried shutting down the agent during that sleep period (also tried a for loop with 20 sleep 1's). There could be a bug here. The agent will wait for health to finish, which will wait for any notifications to finish.

However, shutting down the agent seems to kill the spawn server at once, so basically health seems to wait forever, which in turn the agent will say `Main thread HEALTH takes too long to exit. Giving up...`. And even increasing the time the agent waits for threads to exit to e.g. 30 seconds doesn't help.

However, even after giving up, the agent won't exit, and I'm not sure if it's my setup or something else... We maybe need to check, I've seen this behavior of not shutting down again...

File: health/notifications/alarm-notify.sh.in [hunk 3]
```diff
@@ -1335,21 +1335,33 @@ send_telegram() {
...
-        sent=$((sent + 1))
-      elif [ "${httpcode}" = "401" ]; then
-        error "failed to send telegram notification for: ${host} ${chart}.${name} is ${status} to '${chatid}': Wrong bot token."
-      else
-        error "failed to send telegram notification for: ${host} ${chart}.${name} is ${status} to '${chatid}' with HTTP response status code ${httpcode}."
-      fi
+      successful=0
+      attempt_counter=0
+
+      while [ ${successful} -eq 0 ] && [ ${attempt_counter} -lt 60 ]; do
```
Comment (MrZammler) [inline health/notifications/alarm-notify.sh.in:1341]: Hi @MAH69IK !

While we are investigating the cause of the problem during shutdown, is it possible please to make the `60` configurable via `health_alarm_notify.conf`, and set it as default with 0 retries?

In other words, if someone does not change this value, to work as before with no retries ?

Thanks!

**Comment by ilyam8** [comment]
@MAH69IK, hi. Are you still interested in this PR? If so, we need to address @MrZammler [suggestion](https://github.com/netdata/netdata/pull/13119#discussion_r911822983).

File: health/notifications/alarm-notify.sh.in (hunk 3 above)
Comment (ilyam8) [reply health/notifications/alarm-notify.sh.in]: I went forward and pushed a commit that makes the number of retries configurable and 0 by default. @thiagoftsm can you test it?

**Review by thiagoftsm — DISMISSED** [review]
I tested the PR is everything is working as expected. LGTM!

**Review by MrZammler — APPROVED** [review]
Tested, looks ok!

**Comment by ilyam8** [comment]
@thiagoftsm @MrZammler sorry for the push, but I had to do it 😄 Please, review again.

**Review by MrZammler — APPROVED** [review]

**Review by thiagoftsm — APPROVED** [review]
@ilyam8  LGTM again! :smile:

**Pull request merged**